module/zfs/zio.c

```diff
@@ -2519,8 +2518,7 @@ zio_ddt_collision(zio_t *zio, ddt_t *ddt, ddt_entry_t *dde)
 			    &aflags, &zio->io_bookmark);
 
 			if (error == 0) {
-				if (arc_buf_size(abuf) != zio->io_orig_size ||
```

Comment:
I don't think checking the size is purely an optimization, so you probably still need to do that here. If the sizes are different but the buffers have the same content up to `zio->io_orig_size` -- which is admittedly unlikely given that there must have been a hash collision as well -- the new version wouldn't detect the collision correctly.